CI build improvements

wjordan opened the pull request:
Two improvements to our CI build, one to help debugging and the other for autoscaling frontend readiness:
1. Add extra debug line to `failed to upgrade, removing from rotation` message when a frontend upgrade throws an exception, displaying the error log and a command that can be run to retrieve the log stored on the daemon server.
2. Override the `CDO.app_servers` method to use `knife search` to fetch a dynamic list of Chef nodes containing the `front-end` role in the current environment. As long as we keep our list of active `front-end` nodes up-to-date with the right shutdown hooks (which are currently in place), this lets us upgrade a dynamic list of autoscaled instances from our CI build without having to manually keep the `node['cdo-secrets']['app_servers']` Chef attribute up to date.

Since this change makes the initial `chef-client` run in `ci_build` no longer necessary, I've removed this command for simplicity (`chef-client` is run on the daemon server via `rake build` later on in the build process, so this stops `chef-client` from running twice).

Any remaining static `app_servers` entries still provided via the Chef attribute will be merged into the dynamic list- e.g., this allows us to include a hard-coded `console` entry along with autoscaled instances in the list of front-end servers to be updated.

philbogle commented on deployment.rb:
```diff
@@ -285,6 +285,20 @@ def method_missing(*args)
+    return super unless CDO.chef_managed
+    require 'cdo/rake_utils'
+    servers = RakeUtils.with_bundle_dir(cookbooks_dir) do
+      knife_cmd = "knife search node 'roles:front-end AND chef_environment:#{rack_env}' --format json --attribute cloud_v2"
```
Could you refactors this up as a script that developers can run themselves to see the current list of production front-ends independently of deployment.  I think developers might otherwise find it hard to know exactly what the frontends are now.

philbogle replied:
What will the impact be if chef is unavailable?   Do we need to cache the results of this knife search to handle periods

philbogle replied:
I was wondering whether we needed to check the return result from knife command.  However, I think JSON.parse would always fail in the event the knife command failed, so I think this isn't necessary.  (Empty string is not valid json)

wjordan replied:
Yeah, I think correct behavior if chef is currently unavailable would be toraise the exception rather than return cached results, since the deploy would fail further down the process if chef is unavailable anyway.

philbogle commented:
add_frontend_server still updates the Chef CDO.app_servers attribute.  Just verifying that this is OK-- I guess it is, because the manually scaled servers are not subject to auto-scaling and will get mixed in with the auto-scaled ones.

philbogle commented:
LGTM after adding/updating a standalone script to display the current app servers in production.

wjordan commented:
Actually, the manually scaled servers would be included in the dynamic
result as long as they use the `front-end` role, so we could/should remove
the update from the add script (even if it will be short-lived until
auto-scaling is live on production). I will make a commit to update this.
On May 4, 2016 12:38 PM, "philbogle" notifications@github.com wrote:

> add_frontend_server still updates the Chef CDO.app_servers attribute. Just
> verifying that this is OK-- I guess it is, because the manually scaled
> servers are not subject to auto-scaling and will get mixed in with the
> auto-scaled ones.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> https://github.com/code-dot-org/code-dot-org/pull/8219#issuecomment-216977598

wjordan commented:
Thanks for the review Phil. I've updated the (existing) `ls_frontend_servers` script to list the current production frontend servers using this routine.

After further thought, I will need to make two extra infrastructure changes before this PR can be merged - currently, the daemon-CI server (which is currently in EC2-Classic) does not have SSH access to the autoscaling-managed frontends (which are launched in EC2-VPC, and will be included in the `CDO.app_servers` output once this PR is merged). I'll need to hook up [ClassicLink](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/vpc-classiclink.html) between the EC2-Classic and the VPC, then set up the SSH security group permissions appropriately so the deploy can succeed. I will update this comment when I make progress on this.

Update: I've added a ClassicLink connection between the existing `gateway` and `production-daemon` instances and the VPC where auto scaled instances will live, and the necessary SSH connections seem to be working now.

wjordan commented:
commit eec2ce1 removes the code in the `add_frontend_server` script that manually updates the `CDO.app_servers` list.
